### PR TITLE
[UNR-2093] Upgrade to SpatialOS 14.1.0

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Upgrade the Unreal GDK to use the SpatialOS 14.1.0 worker SDKs.

#### Release note
Upgraded to SpatialOS 14.1.0.

#### Tests
Tests conducted: building and running a local deployment on the UnrealGDKExampleProject. Running a cloud deployment for the same (includes linux build). Standard CI Windows+Linux build.

Ideally, we would also test Mac, iOS, and PS4 builds, but this is probably not worth the effort until we get automatic CI for those.

#### Documentation
A breaking change note along with a release note in the changelog.

#### Primary reviewers
@joshuahuburn @m-samiec @samiwh

### CODE REVIEW

**Comment by mironec** [comment]
Related PRs: 
UnrealGDKExampleProject: https://github.com/spatialos/UnrealGDKExampleProject/pull/58
UnrealEngine: https://github.com/improbableio/UnrealEngine/pull/259

File: Setup.bat [hunk 1]
```diff
@@ -88,25 +88,25 @@ call :MarkEndOfBlock "Create folders"
...
                         "Expand-Archive -Path \"%CORE_SDK_DIR%\worker_sdk\csharp.zip\"                                     -DestinationPath \"%BINARIES_DIR%\Programs\worker_sdk\csharp\" -Force; "^
-                        "Expand-Archive -Path \"%CORE_SDK_DIR%\worker_sdk\c-static-fullylinked-arm-clang_libcpp-ios.zip\"  -DestinationPath \"%BINARIES_DIR%\IOS\" -Force;"^
+                        "Expand-Archive -Path \"%CORE_SDK_DIR%\worker_sdk\c-static-fullylinked-arm-clang-ios.zip\"         -DestinationPath \"%BINARIES_DIR%\IOS\" -Force;"^
                         "Expand-Archive -Path \"%CORE_SDK_DIR%\tools\schema_compiler-x86_64-win32.zip\"                    -DestinationPath \"%BINARIES_DIR%\Programs\" -Force; "^
                         "Expand-Archive -Path \"%CORE_SDK_DIR%\schema\standard_library.zip\"                               -DestinationPath \"%BINARIES_DIR%\Programs\schema\" -Force;"
-    xcopy /s /i /q "%BINARIES_DIR%\Win64\include" "%WORKER_SDK_DIR%"
+    xcopy /s /i /q "%BINARIES_DIR%\Headers\include" "%WORKER_SDK_DIR%"
```
Comment (mironec) [inline Setup.bat]: This could probably be done in a nicer way now. Not sure how we want it.

Comment (samiwh) [reply Setup.bat]: I think you can add include paths to modules

Comment (samiwh) [reply Setup.bat]: Also our worker sdk directory adds extra directories to the worker sdk headers that look odd.

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/WorkerCoordinator/CoordinatorConnection.cs [hunk 2]
```diff
@@ -47,32 +47,20 @@ private static void KeepConnectionAlive(Connection connection, Logger logger)
...
-                using (var dispatcher = new Dispatcher())
-                {
-                    var isConnected = true;
-
-                    dispatcher.OnDisconnect(op =>
-                    {
-                        logger.WriteError("[disconnect] " + op.Reason, logToConnectionIfExists: false);
-                        isConnected = false;
-                    });
+                Dispatcher dispatcher = new Dispatcher();
```
Comment (mironec) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/WorkerCoordinator/CoordinatorConnection.cs]: I have no idea how C# works, but Dispatcher is no longer IDisposable.

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/WorkerCoordinator/CoordinatorConnection.cs [hunk 3]
```diff
@@ -47,32 +47,20 @@ private static void KeepConnectionAlive(Connection connection, Logger logger)
...
+                var isConnected = true;
 
-                    dispatcher.OnLogMessage(op =>
```
Comment (mironec) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/WorkerCoordinator/CoordinatorConnection.cs]: LogMessage deprecated.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp [hunk 4]
```diff
@@ -242,12 +242,16 @@ void USpatialWorkerConnection::ConnectToLocator()
 	FTCHARToUTF8 PlayerIdentityTokenCStr(*LocatorConfig.PlayerIdentityToken);
 	FTCHARToUTF8 LoginTokenCStr(*LocatorConfig.LoginToken);
 
-	Worker_Alpha_LocatorParameters LocatorParams = {};
+	Worker_LocatorParameters LocatorParams = {};
+	FString ProjectName;
+	FParse::Value(FCommandLine::Get(), TEXT("projectName"), ProjectName);
+	LocatorParams.project_name = TCHAR_TO_UTF8(*ProjectName);
+	LocatorParams.credentials_type = Worker_LocatorCredentialsTypes::WORKER_LOCATOR_PLAYER_IDENTITY_CREDENTIALS;
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: Could not find much documentation on this, but I think we have to specify the credential types we are using now.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp [hunk 5]
```diff
@@ -242,12 +242,16 @@ void USpatialWorkerConnection::ConnectToLocator()
 	FTCHARToUTF8 PlayerIdentityTokenCStr(*LocatorConfig.PlayerIdentityToken);
 	FTCHARToUTF8 LoginTokenCStr(*LocatorConfig.LoginToken);
 
-	Worker_Alpha_LocatorParameters LocatorParams = {};
+	Worker_LocatorParameters LocatorParams = {};
+	FString ProjectName;
+	FParse::Value(FCommandLine::Get(), TEXT("projectName"), ProjectName);
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: Gets the project name from the command line, since it's available there. Is there a connection flow that uses the locator that does not get the project name in the command line?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp [hunk 6]
```diff
@@ -61,7 +61,8 @@ void USpatialPlayerSpawner::ReceivePlayerSpawnRequest(Schema_Object* Payload, co
 	// Send a successful response if the player has been accepted, either from this request or one in the past.
 	Worker_CommandResponse CommandResponse = {};
 	CommandResponse.component_id = SpatialConstants::PLAYER_SPAWNER_COMPONENT_ID;
-	CommandResponse.schema_type = Schema_CreateCommandResponse(SpatialConstants::PLAYER_SPAWNER_COMPONENT_ID, 1);
+	CommandResponse.command_index = 1;
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp]: Not sure what this 1 is. Maybe create a SpatialConstants::SOMETHING?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp [hunk 7]
```diff
@@ -59,7 +59,8 @@ bool CreateSpawnerEntity(Worker_SnapshotOutputStream* OutputStream)
 	SpawnerEntity.component_count = Components.Num();
 	SpawnerEntity.components = Components.GetData();
 
-	return Worker_SnapshotOutputStream_WriteEntity(OutputStream, &SpawnerEntity) != 0;
+	Worker_SnapshotOutputStream_WriteEntity(OutputStream, &SpawnerEntity);
+	return Worker_SnapshotOutputStream_GetState(OutputStream).stream_state == WORKER_STREAM_STATE_GOOD;
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: So I am actually not sure how this previously worked. Did 0 indicate success or failure? I return true on a good state, and false on any other, which seems to work.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/OutgoingMessages.cpp [hunk 8]
```diff
@@ -11,28 +11,28 @@ void FEntityQueryRequest::TraverseConstraint(Worker_Constraint* Constraint)
 	{
 	case WORKER_CONSTRAINT_TYPE_AND:
 	{
-		TUniquePtr<Worker_Constraint[]> NewConstraints = MakeUnique<Worker_Constraint[]>(Constraint->and_constraint.constraint_count);
+		TUniquePtr<Worker_Constraint[]> NewConstraints = MakeUnique<Worker_Constraint[]>(Constraint->constraint.and_constraint.constraint_count);
 
-		for (unsigned int i = 0; i < Constraint->and_constraint.constraint_count; i++)
+		for (unsigned int i = 0; i < Constraint->constraint.and_constraint.constraint_count; i++)
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/OutgoingMessages.cpp]: not part of this pr but we shouldn't be using `unsigned int` here. If we want to match the types we should use the fixed width ones.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/OutgoingMessages.cpp]: Actually this isn't worth fixing as this file is deleted in the worker wrapper stuff.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SnapshotManager.cpp [hunk 9]
```diff
@@ -143,7 +143,8 @@ void USnapshotManager::LoadSnapshot(const FString& SnapshotName)
 				// Entity component data must be deep copied so that it can be used for CreateEntityRequest.
 				Schema_ComponentData* CopySchemaData = DeepCopyComponentData(EntityToSpawn->components[i].schema_type);
 				Worker_ComponentData EntityComponentData{};
-				EntityComponentData.component_id = Schema_GetComponentDataComponentId(CopySchemaData);
+				//TODO: Check if this works
+				EntityComponentData.component_id = EntityToSpawn->components[i].component_id;
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SnapshotManager.cpp]: Can't see why this wouldn't work although you may as well store a reference to the worker component object.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialWorkerFlags.h [hunk 10]
```diff
@@ -31,8 +32,9 @@ class SPATIALGDK_API USpatialWorkerFlags : public UBlueprintFunctionLibrary
 	static void UnbindFromOnWorkerFlagsUpdated(const FOnWorkerFlagsUpdatedBP& InDelegate);
 
 	static FOnWorkerFlagsUpdated OnWorkerFlagsUpdated;
+
+	static void ApplyWorkerFlagUpdate(const Worker_FlagUpdateOp& Op);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialWorkerFlags.h]: We should remove the friend if changing the access modifier.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialWorkerFlags.h]: I decided to move it back, didn't notice the friend below at first.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialWorkerFlags.h]: However is this meant to be part of the public interface. Generally this seems like the kind of thing that can be reasonably written without using friends but do we want this to be public?

File: SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h [hunk 11]
```diff
@@ -3,24 +3,23 @@
...
-
 class USpatialPackageMapClient;
 
 struct SPATIALGDK_API FUnrealObjectRef
 {
 	FUnrealObjectRef() = default;
 	FUnrealObjectRef(const FUnrealObjectRef&) = default;
 
-	FUnrealObjectRef(Worker_EntityId Entity, uint32 Offset)
+	FUnrealObjectRef(Worker_EntityId_Key Entity, uint32 Offset)
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h]: Why has this changed? Entity ID widths haven't changed and this doesn't seem like it;s using a map.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h [hunk 12]
```diff
@@ -243,14 +243,14 @@ inline void DeepCopySchemaObject(Schema_Object* Source, Schema_Object* Target)
...
-	Schema_WriteToBuffer(Source, Buffer);
+	Schema_SerializeToBuffer(Source, Buffer, Length);
 	Schema_Clear(Target);
 	Schema_MergeFromBuffer(Target, Buffer, Length);
 }
 
 inline Schema_ComponentData* DeepCopyComponentData(Schema_ComponentData* Source)
 {
-	Schema_ComponentData* Copy = Schema_CreateComponentData(Schema_GetComponentDataComponentId(Source));
+	Schema_ComponentData* Copy = Schema_CreateComponentData();
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h]: I think this function can be rewritten to just use the schema copy function.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h]: I don't think we ever need to serialise stuff ourselves.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp (hunk 7 above)
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: We might want to log on this failing.

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs [hunk 13]
```diff
@@ -43,7 +43,7 @@ public SpatialGDK(ReadOnlyTargetRules Target) : base(Target)
 
         var WorkerLibraryDir = Path.GetFullPath(Path.Combine(ModuleDirectory, "..", "..", "Binaries", "ThirdParty", "Improbable", Target.Platform.ToString()));
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs]: I think we can use this to add an include path to the C headers, which should be easier now that we download the headers and the binaries separately. Not essential but would be nicer.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp (hunk 7 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: This has a corresponding log in SpatialGDKEditorSnapshotGenerator::FillSnapshot, but maybe could be refactored.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp (hunk 6 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp]: It looks like the command index so something like PLAYER_SPAWNER_COMMAND_INDEX or whatever the command is called.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp (hunk 7 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: 0 is implicitly converted to false. Everything else is true.

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/WorkerCoordinator/CoordinatorConnection.cs (hunk 2 above)
Comment (samiwh) [reply SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/WorkerCoordinator/CoordinatorConnection.cs]: Can't find the change in the release note, but if it's not disposable then this seems correct to me.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp (hunk 4 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: The function we want to be using here is `Worker_Locator_ConnectAndQueueAsync`. See the release notes about how `Worker_Alpha_Locator_ConnectAsync` has changed.

File: SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h (hunk 11 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h]: It gave me an error once, but I reran the build with these Key changes reverted, and it works now. The original error was probably a false alarm caused by a change elsewhere, I will revert these now.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: Discussed offline, looked at docs. I am reasonably sure we are maybe using it correctly.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp (hunk 6 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp]: The entire class has hardcoded constants in there, for now I will leave it as is, but may be worth looking into in some future refactoring work.

File: SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h [hunk 14]
```diff
@@ -3,11 +3,10 @@
 #include "CoreMinimal.h"
 
 #include "Utils/SchemaOption.h"
+#include "SpatialCommonTypes.h"
 
 #include <cstdint>
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h]:
```suggestion
```

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h]: I am going to remove SpatialCommonTypes and cstdint and include c_worker.h.

File: Setup.sh [hunk 15]
```diff
@@ -94,27 +94,27 @@ markEndOfBlock "Create folders"
 markStartOfBlock "Retrieve dependencies"
```
Comment (m-samiec) [inline Setup.sh]: Thanks for updating this file also - @jessicafalk thanks you :)

Comment (jessicafalk) [reply Setup.sh]: oh yes, you have no idea how happy this makes me :D

Comment (mironec) [reply Setup.sh]: I must say this file is untested, though.
